roles/tests-integration/tests/common/sniffer.rs

```diff
@@ -0,0 +1,403 @@
+                CoinbaseOutputDataSize(m) => {
+                    AnyMessage::TemplateDistribution(CoinbaseOutputDataSize(m.into_static()))
+                }
+                TemplateDistribution::NewTemplate(m) => AnyMessage::TemplateDistribution(
+                    TemplateDistribution::NewTemplate(m.into_static()),
+                ),
+                TemplateDistribution::RequestTransactionData(m) => {
+                    AnyMessage::TemplateDistribution(TemplateDistribution::RequestTransactionData(
+                        m.into_static(),
+                    ))
+                }
+                TemplateDistribution::RequestTransactionDataError(m) => {
+                    AnyMessage::TemplateDistribution(
+                        TemplateDistribution::RequestTransactionDataError(m.into_static()),
+                    )
+                }
+                TemplateDistribution::RequestTransactionDataSuccess(m) => {
+                    AnyMessage::TemplateDistribution(
+                        TemplateDistribution::RequestTransactionDataSuccess(m.into_static()),
+                    )
+                }
+                TemplateDistribution::SetNewPrevHash(m) => AnyMessage::TemplateDistribution(
+                    TemplateDistribution::SetNewPrevHash(m.into_static()),
+                ),
+                TemplateDistribution::SubmitSolution(m) => AnyMessage::TemplateDistribution(
+                    TemplateDistribution::SubmitSolution(m.into_static()),
+                ),
+            },
+        }
+    }
+
+    async fn wait_for_client(client: SocketAddr) -> TcpStream {
+        let listner = TcpListener::bind(client)
+            .await
+            .expect("Impossible to listen on given address");
+        if let Ok((stream, _)) = listner.accept().await {
+            stream
+        } else {
+            panic!("Impossible to accept dowsntream connetion")
+        }
+    }
+}
+
+// This implementation is used in order to check if a test has handled all messages sent by the
+// downstream and upstream roles. If not, the test will panic.
+//
+// This is useful to ensure that the test has checked all exchanged messages between the roles.
+impl Drop for Sniffer {
+    fn drop(&mut self) {
+        // Dont print backtrace on panic
+        std::panic::set_hook(Box::new(|_| {
+            println!();
+        }));
+        if !self.downstream_messages.is_empty() {
+            println!(
+                "You didnt handle all downstream messages: {:?}",
+                self.downstream_messages
+            );
+            panic!();
+        }
+        if !self.upstream_messages.is_empty() {
+            println!(
+                "You didnt handle all upstream messages: {:?}",
+                self.upstream_messages
+            );
+            panic!();
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct MessagesAggregator {
+    messages: Arc<Mutex<VecDeque<(MsgType, AnyMessage<'static>)>>>,
```

Comment:
@Shourya742 what is the motivation for this change?

Reply:
This can be ignored for now and my next comment is about adding a top. We may need these changes later when the aggregators become read-intensive, but unless we move to a real-world scenario, making theoretical changes doesn’t make sense.